parlai/core/metrics.py

```diff
@@ -115,39 +147,157 @@ class AverageMetric(Metric):
     """
 
-    def remove_articles(text):
-        return re_art.sub(' ', text)
+    @staticmethod
+    def _prec_recall_f1_score(pred_items, gold_items):
+        """
+        Compute precision, recall and f1 given a set of gold and prediction items.
+
+        :param pred_items: iterable of predicted values
+        :param gold_items: iterable of gold values
+
+        :return: tuple (p, r, f1) for precision, recall, f1
+        """
+        common = Counter(gold_items) & Counter(pred_items)
+        num_same = sum(common.values())
+        if num_same == 0:
+            return 0, 0, 0
+        precision = 1.0 * num_same / len(pred_items)
+        recall = 1.0 * num_same / len(gold_items)
+        f1 = (2 * precision * recall) / (precision + recall)
+        return precision, recall, f1
+
+    @staticmethod
+    def compute(guess: str, answers: List[str]) -> 'F1Metric':
+        if guess is None or answers is None:
+            return AverageMetric(0, 0)
+        g_tokens = normalize_answer(guess).split()
+        scores = [
+            F1Metric._prec_recall_f1_score(g_tokens, normalize_answer(a).split())
+            for a in answers
+        ]
+        return F1Metric(max(f1 for p, r, f1 in scores), 1)
+
+
+class ExactMatchMetric(AverageMetric):
+    @staticmethod
+    def compute(guess: str, answers: List[str]) -> 'ExactMatchMetric':
+        if guess is None or answers is None:
+            return None
+        guess = normalize_answer(guess)
+        for a in answers:
+            if guess == normalize_answer(a):
+                return ExactMatchMetric(1)
+        return ExactMatchMetric(0)
+
+
+class BleuMetric(AverageMetric):
+    @staticmethod
+    def compute(guess: str, answers: List[str], k: int = 4) -> Optional['BleuMetric']:
+        """
+        Compute approximate BLEU score between guess and a set of answers.
+        """
+        if nltkbleu is None:
+            # bleu library not installed, just return a default value
+            return None
+        # Warning: BLEU calculation *should* include proper tokenization and
+        # punctuation etc. We're using the normalize_answer for everything though,
+        # so we're over-estimating our BLEU scores.  Also note that NLTK's bleu is
+        # going to be slower than fairseq's (which is written in C), but fairseq's
+        # requires that everything be in arrays of ints (i.e. as tensors). NLTK's
+        # works with strings, which is better suited for this module.
+        weights = [1 / k for _ in range(k)]
+        score = nltkbleu.sentence_bleu(
+            [normalize_answer(a).split(" ") for a in answers],
+            normalize_answer(guess).split(" "),
+            smoothing_function=nltkbleu.SmoothingFunction(epsilon=1e-12).method1,
+            weights=weights,
+        )
+        return BleuMetric(score)
+
 
-    def white_space_fix(text):
-        return ' '.join(text.split())
+class RougeMetric(AverageMetric):
+    _evaluator = None
```

Comment:
what does this do

Reply:
Are you asking about the terrible diff formatting, or the `_evaluator`? It keeps the state of the Rouge calculator in memory (which rouges we're computing, the tokenizer, [some other stuff](https://github.com/Diego999/py-rouge/blob/16f225b1f46e9d382f1bf5170546da218ee98003/rouge/rouge.py#L85-L89)... )